dimod/higherorder/utils.py

```diff
@@ -134,87 +152,77 @@ def make_quadratic(poly, strength, vartype=None, bqm=None):
-                assert len(interaction) >= 2
-                bqm.add_variable(p, bias)
+    poly = BinaryPolynomial(poly, vartype=bqm.vartype)
+    variables = set().union(*poly)
+
+    while any(len(term) > 2 for term in poly):
+        # determine which pair of variables appear most often
+        paircounter = Counter()
+        for term in poly:
+            if len(term) <= 2:
+                # we could leave these in but it can lead to cases like
+                # {'ab': -1, 'cdef': 1} where ab keeps being chosen for
+                # elimination. So we just ignore all the pairs
                 continue
-
-            interaction = tuple(s for s in interaction if s not in pair)
-            interaction += (p,)
-
-        new_poly[interaction] = bias
-
-    return _reduce_degree(bqm, new_poly, vartype, scale)
+            for u, v in itertools.combinations(term, 2):
+                pair = frozenset((u, v))  # so order invarient
+                paircounter[pair] += 1
+        pair, __ = paircounter.most_common(1)[0]
+        u, v = pair
+
+        # make a new product variable p == u*v and replace all (u, v) with p
+        p = _new_product(variables, u, v)
+        terms = [term for term in poly if u in term and v in term]
+        for term in terms:
+            new = tuple(w for w in term if w != u and w != v) + (p,)
+            poly[new] = poly.pop(term)
+
+        # add a constraint enforcing the relationship between p == u*v
+        if vartype is Vartype.BINARY:
+            constraint = _binary_product([u, v, p])
+
+            bqm.info['reduction'][(u, v)] = {'product': p}
+        elif vartype is Vartype.SPIN:
+            aux = _new_aux(variables, u, v)  # need an aux in SPIN-space
+
+            constraint = _spin_product([u, v, p, aux])
+
+            bqm.info['reduction'][(u, v)] = {'product': p, 'auxiliary': aux}
+        else:
+            raise RuntimeError("unknown vartype: {!r}".format(vartype))
+
+        # scale constraint and update the polynomial with it
+        constraint.scale(strength)
+        for v, bias in constraint.linear.items():
+            try:
+                poly[v,] += bias
+            except KeyError:
+                poly[v,] = bias
+        for uv, bias in constraint.quadratic.items():
+            try:
+                poly[uv] += bias
+            except KeyError:
+                poly[uv] = bias
+        try:
+            poly[()] += constraint.offset
+        except KeyError:
+            poly[()] = constraint.offset
+
+    # convert poly to a bqm (it already is one)
+    for term, bias in poly.items():
+        if len(term) == 2:
+            u, v = term
+            bqm.add_interaction(u, v, bias)
+        elif len(term) == 1:
+            v, = term
+            bqm.add_variable(v, bias)
+        elif len(term) == 0:
+            bqm.add_offset(bias)
+        else:
+            # still has higher order terms, this shouldn't happen
+            raise RuntimeError
```

Comment:
can we add a better error message here.

Reply:
Hmm, this is an "internal error", but suppose I could say that the user should file a bug report.

Reply:
yeah it is internal - but if it happens its a legit bug and will help with bug reporting.

Reply:
Also will save frustration for users.